# BoolODE
Git Repo for converting Boolean models to ODE models and performing stochastic simulations.

# Usage
`python src/BoolODE.py --path=data/variables.txt --max-time=200 --num-timepoints=10 --num-experiments=3`

# Options 
```
Usage: BoolODE.py [options]

Options:
  -h, --help            show this help message and exit
  --path=PATH           Path to boolean model file. This file is required.
  --outPrefix=OUTPREFIX
                        [Optional] Prefix for output files. Will create a directories
                        if they are missing.
  --max-time=MAX_TIME   [Optional] Total time of simulation. (Default = 20)
  --num-timepoints=NUM_TIMEPOINTS
                        [Optional] Number of time points to sample. (Default = 100)
  --num-experiments=NUM_EXPERIMENTS
                        [Optional] Number of experiments to perform. (Default = 30)
  -n, --normalize-trajectory
                        [Optional] Min-Max normalize genes across all experiments
  -i, --identical-pars  [Optional] Set single value to similar parameters NOTE: Consider
                        setting this if you are using --sample-pars.
  -s, --sample-pars     [Optional] Sample rate parameters around the hardcoded means ,
                        using 10% stand. dev.
  --write-protein       [Optional] Write both protein and RNA values to file. Useful for
                        debugging.
  -b, --burn-in         [Optional] Treats the first 25% of the time course as burnin ,
                        samples from the rest.
  --inputs=INPUTS       [Optional] Path to input parameter files
  --ics=ICS             [Optional] Path to list of initial conditions
  --strengths=STRENGTHS
                        [Optional] Path to list of interaction strengths
  --species-type=SPECIES_TYPE
                        [Optional] Path to list of molecular species type file. Useful to
                        specify proteins/genes. 
```
# Inputs
BoolODE requires a tab separated file containing a Boolean model. The following is a sample input file
```
Gene	Rule
a	not(c)
c	not(b)
b	not(a)
```
The first column specifies the Gene, while the second column specifies the Boolean rule governing 
the expression of the gene. The rules should be formatted as follows: 
`(a_1 or a_2 ... ) and not (r_1 or r_2 ... )` where `a_`s are the activators, and `r_`s are the
repressors of a given gene.

# Outputs
BoolODE produces the following output files:
1. `parameters.txt` - list of parameters and the corresponding values generated by BoolODE
2. `refNetwork.csv` - An edgelist with signs of interactions inferred from the model file
3. `PseudoTime.csv` - BoolODE uses simulated time as a proxy for pseudotime. This file is used as input to BEELINE
4. `PseudoTime-dropped.csv` - Same as PseudoTime.csv, but experiments with 'low' overall gene expression are dropped
5. `ExpressionData.csv` - The table of gene expression values per 'cell'. For explanation of the format, see below.
6. `ExpressionData-dropped.csv` - Same as ExpressionData.csv, but experiments with 'low' overall gene expression are dropped.
7. `stoch_experiment.txt` - Unprocessed output from BoolODE. 

Additionally, BoolODE creates a `model.py` file in `src/` with the ODE model to be simulated.

The ExpressionData.csv file has rows corresponding to the genes, and
columns corresponding to the timepoints in each experiment.  For
example, `[E0_0,E0_10,E0_20,E1_0,E1_10,E1_20]` shows two experiments
with 3 timepoints, at times 0,10,20 respectively.

# Overview of method
BoolODE is currently designed for transcription factor regulatory networks, though a protein interaction
network can be specified by using the `--species_type` option pointing to a tab separated file containing
If you use BoolODE in your research, please cite:

Aditya Pratapa, Amogh Jalihal, Jeffrey Law, Aditya Bharadwaj, and T M Murali. Benchmarking algorithms for gene regulatory network inference from single-cell transcriptomic data, bioRxiv (2019). doi.org/10.1101/642926
